### PR TITLE
[InstCombine] Explicitly set disjoint flag when converting xor to or.

### PR DESCRIPTION
**Pull request by topperc**

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp [hunk 1]
```diff
@@ -315,6 +315,7 @@ Value *InstCombinerImpl::SimplifyDemandedUseBits(Value *V, APInt DemandedMask,
     if (DemandedMask.isSubsetOf(RHSKnown.Zero | LHSKnown.Zero)) {
       Instruction *Or =
           BinaryOperator::CreateOr(I->getOperand(0), I->getOperand(1));
+      cast<PossiblyDisjointInst>(Or)->setIsDisjoint(true);
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:318]: I don't think this is quite correct. This is only checking that the demanded bits are disjoint. If there is a conflict in the non-demanded bits, adding the `disjoint` flag is still going to make the result poison.

Comment (topperc) [reply llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp]: I've added a check that the demanded mask is all ones.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp [hunk 2]
```diff
@@ -4462,8 +4462,11 @@ Instruction *InstCombinerImpl::visitXor(BinaryOperator &I) {
   Value *Op0 = I.getOperand(0), *Op1 = I.getOperand(1);
   Value *M;
   if (match(&I, m_c_Xor(m_c_And(m_Not(m_Value(M)), m_Value()),
-                        m_c_And(m_Deferred(M), m_Value()))))
-    return BinaryOperator::CreateOr(Op0, Op1);
+                        m_c_And(m_Deferred(M), m_Value())))) {
+    auto *Or = BinaryOperator::CreateOr(Op0, Op1);
```
[the remark this answers is not in the document]
Comment (topperc) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp]: Done and updated the add->or code to use it too.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**